src/runtime_intrinsics.c

```diff
@@ -0,0 +1,930 @@
+
+// This is in implementation of the Julia intrinsic functions against boxed types
+// excluding the c interface (ccall, cglobal, llvmcall)
+//
+// this file assumes a little-endian processor, although that isn't too hard to fix
+// it also assumes two's complement negative numbers, which might be a bit harder to fix
+//
+// TODO: add half-float support
+
+#include "julia.h"
+#include "julia_internal.h"
+#include "APInt-C.h"
+const unsigned int host_char_bit = 8;
+
+// run time version of box/unbox intrinsic
+DLLEXPORT jl_value_t *jl_reinterpret(jl_value_t *ty, jl_value_t *v)
+{
+    JL_TYPECHK(reinterpret, datatype, ty);
+    if (!jl_is_leaf_type(ty) || !jl_is_bitstype(ty))
+        jl_error("reinterpret: target type not a leaf bitstype");
+    if (!jl_is_bitstype(jl_typeof(v)))
+        jl_error("reinterpret: value not a bitstype");
+    if (jl_datatype_size(jl_typeof(v)) != jl_datatype_size(ty))
+        jl_error("reinterpret: argument size does not match size of target type");
+    if (ty == jl_typeof(v))
+        return v;
+    if (ty == (jl_value_t*)jl_bool_type)
+        return *(uint8_t*)jl_data_ptr(v) & 1 ? jl_true : jl_false;
+    return jl_new_bits(ty, jl_data_ptr(v));
+}
+
+// run time version of pointerref intrinsic (warning: i is not rooted)
+DLLEXPORT jl_value_t *jl_pointerref(jl_value_t *p, jl_value_t *i)
+{
+    JL_TYPECHK(pointerref, pointer, p);
+    JL_TYPECHK(pointerref, long, i);
+    jl_value_t *ety = jl_tparam0(jl_typeof(p));
+    if (ety == (jl_value_t*)jl_any_type) {
+        jl_value_t **pp = (jl_value_t**)(jl_unbox_long(p) + (jl_unbox_long(i)-1)*sizeof(void*));
+        return *pp;
+    }
+    else {
+        if (!jl_is_datatype(ety))
+            jl_error("pointerref: invalid pointer");
+        size_t nb = LLT_ALIGN(jl_datatype_size(ety), ((jl_datatype_t*)ety)->alignment);
+        char *pp = (char*)jl_unbox_long(p) + (jl_unbox_long(i)-1)*nb;
+        return jl_new_bits(ety, pp);
+    }
+}
+
+// run time version of pointerset intrinsic (warning: x is not gc-rooted)
+DLLEXPORT jl_value_t *jl_pointerset(jl_value_t *p, jl_value_t *x, jl_value_t *i)
+{
+    JL_TYPECHK(pointerset, pointer, p);
+    JL_TYPECHK(pointerset, long, i);
+    jl_value_t *ety = jl_tparam0(jl_typeof(p));
+    if (ety == (jl_value_t*)jl_any_type) {
+        jl_value_t **pp = (jl_value_t**)(jl_unbox_long(p) + (jl_unbox_long(i)-1)*sizeof(void*));
+        *pp = x;
+    }
+    else {
+        if (!jl_is_datatype(ety))
+            jl_error("pointerset: invalid pointer");
+        size_t nb = LLT_ALIGN(jl_datatype_size(ety), ((jl_datatype_t*)ety)->alignment);
+        char *pp = (char*)jl_unbox_long(p) + (jl_unbox_long(i)-1)*nb;
+        if (jl_typeof(x) != ety)
+            jl_error("pointerset: type mismatch in assign");
+        jl_assign_bits(pp, x);
+    }
+    return p;
+}
+
+
+static inline unsigned int next_power_of_two(unsigned int val) {
+  /* this function taken from libuv src/unix/core.c */
```

Comment:
I wouldn't consider this a "substantial portion" of libuv, so i don't think it needs one (https://github.com/libuv/libuv/blob/26917ddf7a5279c5d913c12121fe955ba4de91c0/LICENSE#L14-L15)